amqp-impl/src/main/java/io/streamnative/pulsar/handlers/amqp/AmqpMultiBundlesChannel.java

```diff
@@ -0,0 +1,348 @@
+                return null;
+            });
+        }
+    }
+
+    public void close() {
+        closeAllConsumers();
+        closeAllProducers();
+        // TODO need to delete exclusive queues in this channel.
+        setDefaultQueue(null);
+    }
+
+    private void closeAllConsumers() {
+        if (log.isDebugEnabled()) {
+            if (!consumerList.isEmpty()) {
+                log.debug("Unsubscribing all consumers on channel  {}", channelId);
+            } else {
+                log.debug("No consumers to unsubscribe on channel {}", channelId);
+            }
+        }
+        try {
+            consumerList.forEach(consumer -> {
+                try {
+                    consumer.close();
+                } catch (Exception e) {
+                    log.error("Failed to close consumer.", e);
+                }
+            });
+            consumerList.clear();
+        } catch (Exception e) {
+            log.error(e.getMessage());
+        }
+    }
+
+    private void closeAllProducers() {
+        for (Producer<byte[]> producer : producerMap.values()) {
+            try {
+                producer.close();
+            } catch (PulsarClientException e) {
+                log.error("Failed to close producer.", e);
+            }
+        }
+    }
+
+    private AmqpAdmin getAmqpAdmin() {
+        return this.connection.getAmqpBrokerService().getAmqpAdmin();
+    }
+
+    public Producer<byte[]> getProducer(String exchange) throws PulsarServerException {
+        PulsarClient client = connection.getPulsarService().getClient();
+        return producerMap.computeIfAbsent(exchange, k -> {
+            try {
+                return client.newProducer()
+                        .topic(getTopicName(PersistentExchange.TOPIC_PREFIX, exchange))
+                        .enableBatching(false)
+                        .create();
+            } catch (PulsarClientException e) {
+                throw new AoPServiceRuntimeException.ProducerCreationRuntimeException(e);
+            }
+        });
+    }
+
+    public CompletableFuture<AmqpPulsarConsumer> getConsumer(String queue, String consumerTag, boolean autoAck) {
+        PulsarClient client;
+        try {
+            client = connection.getPulsarService().getClient();
+        } catch (PulsarServerException e) {
+            return FutureUtil.failedFuture(e);
+        }
+        CompletableFuture<AmqpPulsarConsumer> consumerFuture = new CompletableFuture<>();
+        client.newConsumer()
+                .topic(getTopicName(PersistentQueue.TOPIC_PREFIX, queue))
+                .subscriptionType(SubscriptionType.Shared)
+                .subscriptionName("AMQP_DEFAULT")
+                .subscriptionInitialPosition(SubscriptionInitialPosition.Earliest)
+                .consumerName(UUID.randomUUID().toString())
+                .receiverQueueSize(getConnection().getAmqpConfig().getAmqpPulsarConsumerQueueSize())
```

Comment:
When calling basicReject basicNack and requeue is true, use consumer.negativeAcknowledge API, its default value is 1 minute, should be manual configuration, such as: negativeAckRedeliveryDelay(0, TimeUnit.MILLISECONDS).

Reply:
Ok, I add TODO label first.

Reply:
fixed